File: src/day1.rs
use std::collections::HashMap;

use eyre::Result;

pub fn run(input: &str) -> u64 {

    let p2 = part2(input).unwrap();

    p2
}

pub fn run_debug(input: &str) -> (u64, u64) {
    let p1 = part1(input).unwrap();
    let p2 = part2(input).unwrap();

    (p1, p2)
}

#[derive(Debug)]
struct LastFirst {
    first: u64,
    last: u64,
}

impl LastFirst {
    pub fn new() -> Self {
        Self {
            first: 0,
            last: 0,
        }
    }

    pub fn add_num(&mut self, num: u64) {
        if self.first == 0 {
            self.first = num
        } else {
            self.last = num
        }
    }

    pub fn produce_final_number(&self) -> u64 {
        let mut num = self.first;

        if self.last == 0 {
            num = (num * 10) + self.first;
        } else {
            num = (num * 10) + self.last;
        }

        num
    }

    pub fn clear(&mut self) {
        self.first = 0;
        self.last = 0;
    }
}

pub fn part1(input: &str) -> Result<u64> {
    let mut first_last = LastFirst::new();
    let mut sum = 0;

    for line in input.lines() {
        first_last.clear();
        let chars = line.chars();

        for c in chars {
            if let Some(n) = c.to_digit(10) {
                first_last.add_num(n as u64);
            } else {
                continue
            }
        }
        
        sum += first_last.produce_final_number();
    }

    Ok(sum)
}

pub fn part2(input: &str) -> Result<u64> {
    let cmp_map = [
        ("one", 1),
        ("two", 2),
        ("three", 3),
        ("four", 4),
        ("five", 5),
        ("six", 6),
        ("seven", 7),
        ("eight", 8),
        ("nine", 9),
    ];

    let mut last_first = LastFirst::new();
    let mut sum = 0;
    
    for line in input.lines() {
        let mut cursor = 0;
        last_first.clear();

        'main: while line.len() != cursor {
            let line = &line[cursor..line.len()];

            for (k, v) in cmp_map.iter() {
                if line.starts_with(k) {
                    last_first.add_num(*v);
                    cursor += k.len() - 1;
                    continue 'main;
                }
            }

            let char = &line.chars().next().unwrap();

            if let Some(n) = char.to_digit(10) {
                last_first.add_num(n as u64);
            }

            cursor += 1;
        }    

        sum += last_first.produce_final_number();
    }

    Ok(sum)
}

File: zig/src/day1.zig
const std = @import("std");
const print = std.debug.print;

pub fn part1(input: []const u8) u64 {
    var lines = std.mem.splitAny(u8, input, "\n");

    var digits = [_]u8{ 0, 0 };
    var sum: u64 = 0;

    while (lines.next()) |line| {
        var digit: u8 = 0;
        for (line) |char| {
            digit = std.fmt.charToDigit(char, 10) catch continue;

            if (digits[0] == 0) {
                digits[0] = digit;
            }
        }

        digits[1] = digit;

        const num = digits[0] * 10 + digits[1];

        sum += num;

        digits[0] = 0;
        digits[1] = 0;
    }

    return sum;
}

const KeyValue = struct { key: []const u8, value: u8 };

pub fn part2(input: []const u8) u64 {
    const values = [_]KeyValue{
        KeyValue{ .key = "one", .value = 1 },
        KeyValue{ .key = "two", .value = 2 },
        KeyValue{ .key = "three", .value = 3 },
        KeyValue{ .key = "four", .value = 4 },
        KeyValue{ .key = "five", .value = 5 },
        KeyValue{ .key = "six", .value = 6 },
        KeyValue{ .key = "seven", .value = 7 },
        KeyValue{ .key = "eight", .value = 8 },
        KeyValue{ .key = "nine", .value = 9 },
    };

    var sum: u64 = 0;

    var first_digit: u8 = 0;
    var last_digit: u8 = 0;

    var lines = std.mem.splitAny(u8, input, "\n");
    while (lines.next()) |line| {
        var n: usize = 0;
        first_digit = 0;
        last_digit = 0;
        char_loop: while (n < line.len) : (n += 1) {
            for (values) |kv| {
                if (std.mem.startsWith(u8, line[n..], kv.key)) {
                    if (first_digit == 0) {
                        first_digit = kv.value;
                    } else {
                        last_digit = kv.value;
                    }

                    n += kv.key.len - 1;
                    continue :char_loop;
                }
            }

            const d = std.fmt.charToDigit(line[n], 10) catch continue;

            if (first_digit == 0) {
                first_digit = d;
            } else {
                last_digit = d;
            }
        }

        if (last_digit == 0) last_digit = first_digit;

        const num = first_digit * 10 + last_digit;

        print("{d}\n", .{num});
        sum += num;
    }

    return sum;
}

pub fn run() !void {
    //var alloc = std.heap.GeneralPurposeAllocator(.{}){};

    //const buff = try std.fs.cwd()
    //.readFileAlloc(alloc.allocator(), "inputs/day1_input.txt", std.math.maxInt(usize));
    const buff = @embedFile("./inputs/day1_test.txt");

    var timer = try std.time.Timer.start();
    const res_p1 = part1(buff);
    timer.reset();

    const p1_time: f64 = @as(f64, @floatFromInt(timer.read())) / 1e6;

    const res_p2 = part2(buff);

    std.debug.print("Results: {d} ({d}ms) | {d}\n", .{ res_p1, p1_time, res_p2 });
}
